ukvm/ukvm_hv_openbsd_x86_64.c

```diff
@@ -0,0 +1,254 @@
+
+    if ((pw = getpwnam(VMD_USER)) == NULL)
+        err(1, "can't get _vmd user");
+    uid = pw->pw_uid;
+    gid = pw->pw_gid;
+
+    if (chroot(pw->pw_dir) == -1)
+        err(1, "chroot: %s", pw->pw_dir);
+
+    if (chdir("/") == -1)
+        err(1, "chdir(\"/\")");
+
+    if (setgroups(1, &gid) ||
+        setresgid(gid, gid, gid) ||
+        setresuid(uid, uid, uid))
+        err(1, "unable to revoke privs");
+
+    /*
+     * pledge in the vm processes:
+     * stdio - for malloc and basic I/O including events.
+     * vmm - for the vmm ioctls and operations.
+     */
+    if (pledge("stdio vmm", NULL) == -1)
+        err(errno, "pledge");
+
+    for (;;) {
+        vrp->vrp_irq = 0xFFFF;
+
+        if (ioctl(hvb->vmd_fd, VMM_IOC_RUN, vrp) < 0) {
+            if(errno == EIO) // guest halted with interrupts disabled, ie cli, hlt
+                return;
+
+            err(1, "ukvm_hv_vcpu_loop: vm / vcpu run ioctl failed");
+        }
+
+        /* If the VM is terminating, exit normally */
+        if (vrp->vrp_exit_reason == VM_EXIT_TERMINATED) {
+            return;
+        }
+
+        union vm_exit *vei = vrp->vrp_exit;
+
+        if (vrp->vrp_exit_reason != VM_EXIT_NONE) {
+            switch (vrp->vrp_exit_reason) {
+                case VMX_EXIT_INT_WINDOW:
+                case VMX_EXIT_CPUID:
+                case VMX_EXIT_EXTINT:
+                case SVM_VMEXIT_INTR:
+                case VMX_EXIT_EPT_VIOLATION:
+                case SVM_VMEXIT_NPF:
+                case SVM_VMEXIT_MSR:
+                case SVM_VMEXIT_CPUID:
+                    // nothing to be done here, as per vmd
+                    break;
+                case VMX_EXIT_IO:
+                case SVM_VMEXIT_IOIO:
+                    if (vei->vei.vei_dir != VEI_DIR_OUT
+                            || vei->vei.vei_size != 4)
+                        errx(1, "Invalid guest port access: port=0x%x", vei->vei.vei_port);
+                    if (vei->vei.vei_port < UKVM_HYPERCALL_PIO_BASE ||
+                            vei->vei.vei_port >= (UKVM_HYPERCALL_PIO_BASE + UKVM_HYPERCALL_MAX))
+                        errx(1, "Invalid guest port access: port=0x%x", vei->vei.vei_port);
+
+                    int nr = vei->vei.vei_port - UKVM_HYPERCALL_PIO_BASE;
+                    ukvm_hypercall_fn_t fn = ukvm_core_hypercalls[nr];
+                    if (fn == NULL)
+                        errx(1, "Invalid guest hypercall: num=%d", nr);
+
+                    ukvm_gpa_t gpa = vei->vei.vei_data;
+                    fn(hv, gpa);
+                    break;
+                case VMX_EXIT_HLT:
+                case SVM_VMEXIT_HLT:
+                    // stop on exit
+                    break;
+                case VMX_EXIT_TRIPLE_FAULT:
+                case SVM_VMEXIT_SHUTDOWN:
+                    /* reset VM */
+                    errx(1, "Triple Fault");
+                default:
+                    errx(1, "unknown exit reason");
```

Comment:
Use something like the code from FreeBSD so that these can be debugged, otherwise the operator has no idea what happened:
```
default: {
            errx(1, "unhandled exit: exitcode=%d, rip=0x%" PRIx64,
                    vme->exitcode, vme->rip);
        }
```